### ISSUE TITLE
Fails to properly render PDF 

### ISSUE THREAD

**Issue by cpbl** [Cimbali/pympress#159]
# Environment
- Ubuntu 20.04
- Python 3.8.2

# Pympress
- version 1.5.1
- apt

# Expected behaviour 
render my presentation as other PDF viewers do

# Steps to reproduce
 pympress --log=DEBUG testbug.pdf

# Problem
Text is missing from slide

[testbug.pdf](https://github.com/Cimbali/pympress/files/4725749/testbug.pdf)
![pympressbug](https://user-images.githubusercontent.com/3696165/83679816-04b3b480-a5ae-11ea-9161-bc482a721191.png)

No log file is created in any of the following locations:

> /tmp/pympress.log
> /var/tmp/pympress.log
> /usr/tmp/pympress.log
> C:\Users\<username>\AppData\Local\Temp\pympress.log
> C:\TEMP\pympress.log
> C:\TMP\pympress.log

**Comment by Cimbali** [comment]
Interesting, I'll try and have a look soon. The actual log file location is given in the pop-up avaible through the `Help > about` menu item.

**Comment by cpbl** [comment]
Thanks. It would be useful to have that `Help >about` hint in the Issue template. Here's the log file:
```

ERROR:pympress.extras:Video support using GStreamer is disabled.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pympress/extras.py", line 408, in _setup_backends
    from pympress.media_overlays.gst_backend import GstOverlay
  File "/usr/lib/python3/dist-packages/pympress/media_overlays/gst_backend.py", line 36, in <module>
    gi.require_version('GstPlayer', '1.0')
  File "/usr/lib/python3/dist-packages/gi/__init__.py", line 129, in require_version
    raise ValueError('Namespace %s not available' % namespace)
ValueError: Namespace GstPlayer not available
INFO:pympress.__main__:Pympress: 1.5.1 ; Python: 3.8.2 ; OS: Linux 5.4.0-33-generic ; Gtk 3.24.18 ; GLib 2.64.1 ; Poppler 0.86.1 cairo ; Cairo 1.16.0 , pycairo 1.16.2 ; Media: GtkImage gif player, VLC 3.0.9.2 Vetinari
ERROR:pympress.builder:Can not reach target of signal <scribble.Scribbler object at 0x7fe821d93ec0 (pympress+scribble+Scribbler at 0x281e380)>.laser.track_enter_leave()
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pympress/builder.py", line 129, in find_callback_handler
    return getattr(target, handler_name)
AttributeError: 'Scribbler' object has no attribute 'laser.track_enter_leave'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pympress/builder.py", line 109, in signal_resolver
    target = getattr(target, attr)
AttributeError: 'Scribbler' object has no attribute 'laser'
ERROR:pympress.builder:Can not reach target of signal <scribble.Scribbler object at 0x7fe821d93ec0 (pympress+scribble+Scribbler at 0x281e380)>.laser.track_enter_leave()
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pympress/builder.py", line 129, in find_callback_handler
    return getattr(target, handler_name)
AttributeError: 'Scribbler' object has no attribute 'laser.track_enter_leave'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pympress/builder.py", line 109, in signal_resolver
    target = getattr(target, attr)
AttributeError: 'NoneType' object has no attribute 'track_enter_leave'
CRITICAL:pympress.builder:Impossible to connect signal enter-notify-event from object <Gtk.DrawingArea object at 0x7fe820c0ad80 (GtkDrawingArea at 0x2743570)> to handler laser.track_enter_leave
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pympress/builder.py", line 172, in signal_connector
    object.connect(signal_name, handler, *user_data)
TypeError: second argument must be callable
ERROR:pympress.builder:Can not reach target of signal <scribble.Scribbler object at 0x7fe821d93ec0 (pympress+scribble+Scribbler at 0x281e380)>.laser.track_enter_leave()
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pympress/builder.py", line 129, in find_callback_handler
    return getattr(target, handler_name)
AttributeError: 'Scribbler' object has no attribute 'laser.track_enter_leave'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pympress/builder.py", line 109, in signal_resolver
    target = getattr(target, attr)
AttributeError: 'Scribbler' object has no attribute 'laser'
ERROR:pympress.builder:Can not reach target of signal <scribble.Scribbler object at 0x7fe821d93ec0 (pympress+scribble+Scribbler at 0x281e380)>.laser.track_enter_leave()
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pympress/builder.py", line 129, in find_callback_handler
    return getattr(target, handler_name)
AttributeError: 'Scribbler' object has no attribute 'laser.track_enter_leave'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pympress/builder.py", line 109, in signal_resolver
    target = getattr(target, attr)
AttributeError: 'NoneType' object has no attribute 'track_enter_leave'
CRITICAL:pympress.builder:Impossible to connect signal leave-notify-event from object <Gtk.DrawingArea object at 0x7fe820c0ad80 (GtkDrawingArea at 0x2743570)> to handler laser.track_enter_leave
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pympress/builder.py", line 172, in signal_connector
    object.connect(signal_name, handler, *user_data)
TypeError: second argument must be callable

```

**Comment by Cimbali** [comment]
> Thanks. It would be useful to have that `Help >about` hint in the Issue template

Indeed the issue template is not up to date anymore, we found a better place for log files than the temp folders.

---

I see unrelated issues in your log that may be worth looking into separately, but the part with all the software versions is missing. As I can’t reproduce the issue and we’ve had rendering problems before due to the poppler library, I’m suspecting i might be an older Poppler version.

The line in the log I’m looking for is something like this:
> INFO:pympress.__main__:Pympress: 1.5.3 ; Python: 3.8.2 ; OS: Linux 5.6.11-1-default #1 SMP Wed May 6 10:42:09 UTC 2020 (91c024a) ; Gtk 3.24.20 ; GLib 2.64.2 ; Poppler 0.88.0 cairo ; Cairo 1.16.0 , pycairo 1.19.1 ; Media: GtkImage gif player, VLC 3.0.10 Vetinari

**Comment by cpbl** [comment]
@cimbali Isn't that the 10th line in the log file I submitted, above?

**Comment by Cimbali** [comment]
It is! My bad. :)

**Comment by dasfreak** [comment]
Hey,

same issue here:

`INFO:pympress.__main__:Pympress: 1.5.3 ; Python: 3.8.3 ; OS: Linux 5.7.0-3-MANJARO #1 SMP PREEMPT Wed Jun 3 03:15:33 UTC 2020 ; Gtk 3.24.20 ; GLib 2.64.1 ; Poppler 0.88.0 cairo ; Cairo 1.17.3 , pycairo 1.19.1 ; Media: GtkImage gif player, GStreamer 1.16.2, VLC 3.0.10 Vetinari`

However, I was able to circumvent the issue by manually setting the labels alpha to 0 `framealpha=0` [1].

Best regards

[1] https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.legend.html

**Comment by cpbl** [comment]
Is there a chance this is matplotlib's fault? My PDF was also created by matplotlib. Is it possible to made non-standards-compliant PDF? @dasfreak would it be helpful to @Cimbali to provide two small PDFs, with the only difference being the alpha you set?

**Comment by dasfreak** [comment]
Here is a minimal (not) working example.
- `create_figures.py` is used to generate two figures as PDFs.
- `test.tex` will generate a very simple PDF containing both figures

[mwe.zip](https://github.com/Cimbali/pympress/files/4873645/mwe.zip)

```
matplotlib      3.2.2
numpy           1.19.0
pandas          1.0.5
seaborn         0.10.1
```

![image](https://user-images.githubusercontent.com/7197923/86519019-e522dc80-be36-11ea-94fb-0e1c7816447e.png)
Left is pympress (1.5.3), right is evince (3.36.3-1)


If you need more information feel free to contact me.

Best,
dasFreak

**Comment by Cimbali** [comment]
Well it’s strange, I still can’t reproduce these rendering issues. For me they render exactly the same in any pdf reader and in pympress.

Does the `framealpha=0` workaround work for you @cpbl ?

In any case, I’ll try to see how the legend is implemented in the pdf. Maybe it’s an annotation of some kind and we then don’t respect z-ordering or something like that… Not sure how that could happen but I’ll try to dig deeper.

**Comment by cpbl** [comment]
I tried @dasfreak's MWE and it looks the same as for him.
I'm using similar versions of stuff to make the pdfs:

```
import matplotlib,numpy,pandas,seaborn                                  
matplotlib.__version__,numpy.__version__, pandas.__version__, seaborn.__version__                                                               
Out[2]: ('3.2.1', '1.17.4', '1.0.4', '0.10.1')

```
Thanks

**Comment by sebp** [comment]
I'm having the same problem on Fedora 32 using the package from copr.

```
Pympress: 1.5.3 
Python: 3.8.3 
OS: Linux 5.7.7-200.fc32.x86_64 #1 SMP Wed Jul 1 19:53:01 UTC 2020 
Gtk 3.24.20 
GLib 2.64.1 
Poppler 0.84.0 cairo 
Cairo 1.16.0 , pycairo 1.18.2 
Media: GtkImage gif player, GStreamer 1.16.2
```

In addition, I noticed that if plotting things on top of each other, like text on top of a bar, the text disappears. This happens when saving PDFs with `transparent=True` too.

Result with pympress:
![pympress](https://user-images.githubusercontent.com/74460/87351733-64a46000-c549-11ea-9c02-8bf5683d6b80.png)

Result with evince:
![evince](https://user-images.githubusercontent.com/74460/87351752-6bcb6e00-c549-11ea-9e45-3af5bd64e8c5.png)

**Comment by Cimbali** [comment]
So it does not seem to be an annotation issue, however maybe the small modifications related to transparency below can fix it. I can’t reproduce this bug, so if anyone who has the bug can test these changes:

```diff
diff --git a/pympress/document.py b/pympress/document.py
index 07a79c8..837c9fe 100644
--- a/pympress/document.py
+++ b/pympress/document.py
@@ -567,7 +567,7 @@ class Page(object):
         """
         pw, ph = self.get_size(dtype)
 
-        cr.set_source_rgb(1, 1, 1)
+        cr.set_source_rgba(1, 1, 1, 1)
 
         # Scale
         scale = min(ww / pw, wh / ph)
diff --git a/pympress/surfacecache.py b/pympress/surfacecache.py
index b044f23..3e542cc 100644
--- a/pympress/surfacecache.py
+++ b/pympress/surfacecache.py
@@ -294,7 +294,7 @@ class SurfaceCache(object):
 
         # Render to a ImageSurface
         try:
-            surface = self.surface_factory[widget_name](cairo.CONTENT_COLOR, ww, wh)
+            surface = self.surface_factory[widget_name](cairo.CONTENT_COLOR_ALPHA, ww, wh)
         except AttributeError:
             logger.warning('Widget {} was not mapped when rendering'.format(widget_name), exc_info = True)
             return False
```

They are also available at f248005, which you can install with pip using:
```bash
pip install git+https://github.com/Cimbali/pympress.git@f248005b538935d1048abd2890c7a380a3300aea
```

**Comment by sebp** [comment]
Unfortunately, it does not fix it for me, it looks exactly the same.

**Comment by dannosliwcd** [comment]
If I modify pympress to use a Cairo image back end, the issues appear to be resolved (the original note about missing text, as well as the opacity issue).

My configuration is able to reproduce this issue, and has been defaulting to the Xlib renderer in Cairo. I recently tried out the test pdf with the latest release of pympress on Windows, and the issues does not reproduce there. I'm curious which rendering back ends are being used by people in this thread, aside from win32 and xlib.

I'm not sure if there's an easy way to check which back end is the default for a given system. I just set a breakpoint after the call to `create_similar_image_surface()` in ui.py and checked the type of the returned value.
